<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="John Blischak" />

<meta name="date" content="2018-01-26" />

<title>Differential expression analysis for studies with 2 groups</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">DE analysis with limma</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/dc-bioc-limma">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Differential expression analysis for studies with 2 groups</h1>
<h4 class="author"><em>John Blischak</em></h4>
<h4 class="date"><em>2018-01-26</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<p><strong>Last updated:</strong> 2018-02-16</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> 4d75cfb</p>
<p>Analyze proteomics data set that compares 2 groups.</p>
<pre class="r"><code>library(&quot;Biobase&quot;)
library(&quot;ggplot2&quot;)
library(&quot;limma&quot;)</code></pre>
<pre><code>Warning: package &#39;limma&#39; was built under R version 3.3.3</code></pre>
<pre class="r"><code>eset &lt;- readRDS(&quot;../data/ch02.rds&quot;)</code></pre>
<div id="studies-with-2-groups-video" class="section level2">
<h2>Studies with 2 groups (Video)</h2>
<p>Describe the scientific question, the experimental design, and the data collected for the 2-group study. Introduce the ExpressionSet object that contains the data. Review the quality control procedures covered in past Bioconductor courses (specifically comparing distributions and PCA).</p>
</div>
<div id="plot-a-differentially-expressed-gene" class="section level2">
<h2>Plot a differentially expressed gene</h2>
<p>Create boxplots of a few pre-selected genes (one clearly DE, one clearly not, and one in between). Use <code>pData</code> to select the phenotype variables and <code>exprs</code> to access the expression data.</p>
<pre class="r"><code>boxplot(exprs(eset)[3, ] ~ pData(eset)$spikein, main = fData(eset)$protein[3])</code></pre>
<p><img src="figure/ch02.Rmd/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>boxplot(exprs(eset)[5, ] ~ pData(eset)$spikein, main = fData(eset)$protein[5])</code></pre>
<p><img src="figure/ch02.Rmd/unnamed-chunk-1-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>boxplot(exprs(eset)[463, ] ~ pData(eset)$spikein, main = fData(eset)$protein[463])</code></pre>
<p><img src="figure/ch02.Rmd/unnamed-chunk-1-3.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="create-a-density-plot-to-compare-distribution-across-samples" class="section level2">
<h2>Create a density plot to compare distribution across samples</h2>
<p>Use <code>limma::plotDensities</code> to confirm that the distribution of gene expression levels is consistent across the samples.</p>
<pre class="r"><code>plotDensities(exprs(eset), legend = FALSE)</code></pre>
<p><img src="figure/ch02.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="create-a-pca-plot-to-assess-source-of-variation" class="section level2">
<h2>Create a PCA plot to assess source of variation</h2>
<p>Use <code>prcomp</code> to compute principal components and then plot PC1 vs. PC2 to confirm that the biological effect is the main source of variation.</p>
<pre class="r"><code>pca &lt;- prcomp(t(exprs(eset)), scale. = TRUE)
plot(pca)</code></pre>
<p><img src="figure/ch02.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>d &lt;- data.frame(pData(eset), pca$x)
ggplot(d, aes(x = PC1, y = PC2, color = spikein)) +
  geom_point()</code></pre>
<p><img src="figure/ch02.Rmd/unnamed-chunk-3-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="limma-for-differential-expression-video" class="section level2">
<h2>limma for differential expression (Video)</h2>
<p>Describe the standard limma workflow. Describe the 2 main techniques for constructing the linear model: treatment-contrasts versus group-means parametrizations.</p>
</div>
<div id="create-design-matrix-for-treatment-contrasts-parametrization" class="section level2">
<h2>Create design matrix for treatment-contrasts parametrization</h2>
<p>Use <code>model.matrix</code> to create a linear model with an intercept and one binary variable. Use <code>colSums</code> to reason about how this relates to the samples (e.g. the intercept represents the mean across samples because it is 1 for every sample).</p>
<pre class="r"><code>design &lt;- model.matrix(~spikein, data = pData(eset))
colSums(design)</code></pre>
<pre><code>  (Intercept) spikein25fmol 
            6             3 </code></pre>
</div>
<div id="fit-and-test-model-for-treatment-contrasts-parametrization" class="section level2">
<h2>Fit and test model for treatment-contrasts parametrization</h2>
<p>Use <code>limma::lmFit</code>, <code>limma::eBayes</code>, and <code>limma::decideTests</code> to fit and test the model.</p>
<pre class="r"><code>fit &lt;- lmFit(eset, design)
head(fit$coefficients)</code></pre>
<pre><code>  (Intercept) spikein25fmol
1    21.19598   0.134121520
2    32.48541  -0.004568881
3    29.78624   1.559134985
4    25.25850   0.706384928
5    29.30042  -0.009631149
6    22.12669   1.089862564</code></pre>
<pre class="r"><code>fit &lt;- eBayes(fit)
results &lt;- decideTests(fit[, 2])
summary(results)</code></pre>
<pre><code>   spikein25fmol
-1             9
0           1879
1             56</code></pre>
</div>
<div id="create-design-matrix-and-contrasts-matrix-for-group-means-parametrization" class="section level2">
<h2>Create design matrix and contrasts matrix for group-means parametrization</h2>
<p>Use <code>model.matrix</code> to create a linear model with two binary variables (and no intercept). Use <code>colSums</code> to reason about how this relates to the samples (e.g. each of the terms represents the mean for its group of samples because it is the only term that is 1 for those samples). Use <code>limma::makeContrasts</code> to create a contrasts matrix based on this new linear model.</p>
<pre class="r"><code>design &lt;- model.matrix(~0 + spikein, data = pData(eset))
colSums(design)</code></pre>
<pre><code>spikein10fmol spikein25fmol 
            3             3 </code></pre>
<pre class="r"><code>cont_mat &lt;- makeContrasts(spike_effect = spikein25fmol - spikein10fmol,
                          levels = design)
cont_mat</code></pre>
<pre><code>               Contrasts
Levels          spike_effect
  spikein10fmol           -1
  spikein25fmol            1</code></pre>
</div>
<div id="fit-and-test-model-for-group-means-parametrization" class="section level2">
<h2>Fit and test model for group-means parametrization</h2>
<p>Use <code>limma::lmFit</code>, <code>limma::contrasts.fit</code>, <code>limma::eBayes</code>, and <code>limma::decideTests</code> to fit and test the model. Confirm that the results are identical to the more traditional linear modelling approach used previously.</p>
<pre class="r"><code>fit &lt;- lmFit(eset, design)
head(fit$coefficients)</code></pre>
<pre><code>  spikein10fmol spikein25fmol
1      21.19598      21.33010
2      32.48541      32.48084
3      29.78624      31.34538
4      25.25850      25.96488
5      29.30042      29.29079
6      22.12669      23.21656</code></pre>
<pre class="r"><code>fit2 &lt;- contrasts.fit(fit, contrasts = cont_mat)
head(fit2$coefficients)</code></pre>
<pre><code>   Contrasts
    spike_effect
  1  0.134121520
  2 -0.004568881
  3  1.559134985
  4  0.706384928
  5 -0.009631149
  6  1.089862564</code></pre>
<pre class="r"><code>fit2 &lt;- eBayes(fit2)
results &lt;- decideTests(fit2)
summary(results)</code></pre>
<pre><code>   spike_effect
-1            9
0          1879
1            56</code></pre>
</div>
<div id="visualizing-the-results-video" class="section level2">
<h2>Visualizing the results (Video)</h2>
<p>Describe how to access the results with <code>topTable</code> and describe the columns. Demonstrate some common visualizations.</p>
</div>
<div id="create-a-histogram-of-p-values" class="section level2">
<h2>Create a histogram of p-values</h2>
<p>Use <code>geom_histogram</code> to plot <code>P.Value</code> column from <code>limma::topTable</code>. Ask question to confirm they understand that the p-value distribution corresponds to the number of differentially expressed genes identified.</p>
<pre class="r"><code>topTable(fit2)</code></pre>
<pre><code>                                                                  protein
33                           P06396upsedyp|GELS_HUMAN_upsedyp;CON__Q3SX14
34                                       P06732upsedyp|KCRM_HUMAN_upsedyp
29                                       P02787upsedyp|TRFE_HUMAN_upsedyp
53                P68871upsedyp|HBB_HUMAN_upsedyp;CON__Q3SX09;CON__P02070
30               P02788upsedyp|TRFL_HUMAN_upsedyp;CON__Q29443;CON__Q0IIK2
39                                       P10599upsedyp|THIO_HUMAN_upsedyp
15                                      P00709upsedyp|LALBA_HUMAN_upsedyp
42                                       P15559upsedyp|NQO1_HUMAN_upsedyp
11                                      O00762upsedyp|UBE2C_HUMAN_upsedyp
3  P02768upsedyp|ALBU_HUMAN_upsedyp;CON__P02768-1;CON__P02769;CON__Q3SZ57
      logFC  AveExpr        t      P.Value    adj.P.Val        B
33 1.861560 29.47426 27.52384 5.176104e-07 0.0002584916 7.311601
34 1.534331 29.14056 27.01624 5.719561e-07 0.0002584916 7.221913
29 1.596815 30.67280 26.77212 6.004868e-07 0.0002584916 7.177897
53 1.485236 28.84136 25.03177 8.609292e-07 0.0002584916 6.846476
30 1.502014 31.02911 24.79491 9.059136e-07 0.0002584916 6.798845
39 1.473717 28.31979 24.30523 1.008053e-06 0.0002584916 6.698332
15 1.468805 28.42487 24.18606 1.034938e-06 0.0002584916 6.673445
42 1.541838 27.32908 23.91733 1.098739e-06 0.0002584916 6.616701
11 1.554906 28.71986 23.25450 1.277099e-06 0.0002584916 6.472931
3  1.559135 30.56581 23.07978 1.329689e-06 0.0002584916 6.434105</code></pre>
<pre class="r"><code>stats &lt;- topTable(fit2, number = nrow(fit2), sort.by = &quot;none&quot;)
ggplot(stats, aes(x = P.Value)) +
  geom_histogram()</code></pre>
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="figure/ch02.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="create-a-volcano-plot" class="section level2">
<h2>Create a volcano plot</h2>
<p>Use <code>geom_point()</code> to plot <code>-log10(P.Value)</code> vs. <code>logFC</code>. Mention <code>limma::volcanoPlot</code> after exercise is completed.</p>
<pre class="r"><code>stats$significant &lt;- stats$adj.P.Val &lt; 0.05
ggplot(stats, aes(x = logFC, y = -log10(P.Value), color = significant)) +
  geom_point()</code></pre>
<p><img src="figure/ch02.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<!-- Insert the session information into the document -->
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.3.2 (2016-10-31)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X Yosemite 10.10.5

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] limma_3.30.13       ggplot2_2.2.1       Biobase_2.34.0     
[4] BiocGenerics_0.20.0

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.14     knitr_1.18       magrittr_1.5     munsell_0.4.3   
 [5] colorspace_1.3-2 rlang_0.1.6      stringr_1.2.0    plyr_1.8.4      
 [9] tools_3.3.2      grid_3.3.2       gtable_0.2.0     git2r_0.21.0    
[13] htmltools_0.3.6  yaml_2.1.16      lazyeval_0.2.1   rprojroot_1.3-2 
[17] digest_0.6.13    tibble_1.3.3     evaluate_0.10.1  rmarkdown_1.8   
[21] labeling_0.3     stringi_1.1.5    scales_0.4.1     backports_1.1.2 </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
