<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="John Blischak" />

<meta name="date" content="2018-08-20" />

<title>Understanding linear models</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">DE analysis with limma</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/dc-bioc-limma">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Understanding linear models</h1>
<h4 class="author">John Blischak</h4>
<h4 class="date">2018-08-20</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-04-10
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>dc-bioc-limma/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.2.0.9000). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed12345code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(12345)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed12345code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(12345)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjdblischakdcbioclimmatree6f4d83b51640a0b5af65f590f35cfbf00ef45b92targetblank6f4d83ba"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jdblischak/dc-bioc-limma/tree/6f4d83b51640a0b5af65f590f35cfbf00ef45b92" target="_blank">6f4d83b</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjdblischakdcbioclimmatree6f4d83b51640a0b5af65f590f35cfbf00ef45b92targetblank6f4d83ba" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/table-s1.txt
    Untracked:  analysis/table-s2.txt
    Untracked:  code/tb-scratch.R
    Untracked:  data/counts_per_sample.txt
    Untracked:  docs/table-s1.txt
    Untracked:  docs/table-s2.txt
    Untracked:  factorial-dox.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jdblischak/dc-bioc-limma/41b57eef2720b8f405efe2967a15d4a88be89b5e/docs/linear-models.html" target="_blank">41b57ee</a>
</td>
<td>
John Blischak
</td>
<td>
2019-02-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jdblischak/dc-bioc-limma/2372aa1e60c379d881b48e77a44fed980fc69c22/docs/linear-models.html" target="_blank">2372aa1</a>
</td>
<td>
John Blischak
</td>
<td>
2019-01-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jdblischak/dc-bioc-limma/f440a87dab57459b871f06dbb7f728e2d95266cb/docs/linear-models.html" target="_blank">f440a87</a>
</td>
<td>
John Blischak
</td>
<td>
2018-08-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jdblischak/dc-bioc-limma/49764900d3ce4ebe074d21c8736c0acaf1c94e97/docs/linear-models.html" target="_blank">4976490</a>
</td>
<td>
John Blischak
</td>
<td>
2018-08-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/192083341b74a84f8295ca7bf85f48e443865751/analysis/linear-models.Rmd" target="_blank">1920833</a>
</td>
<td>
John Blischak
</td>
<td>
2018-08-20
</td>
<td>
Refactor first edition of chapter 1 into distinct lessons.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The goal is to provide some visualizations to help you understand linear models.</p>
</div>
<div id="setup" class="section level2">
<h2>Setup</h2>
<pre class="r"><code>library(&quot;broom&quot;)
library(&quot;cowplot&quot;)
library(&quot;dplyr&quot;)
library(&quot;ggplot2&quot;)
theme_set(theme_classic(base_size = 16))
library(&quot;knitr&quot;)
opts_chunk$set(fig.width = 10, fig.height = 5, message = FALSE)</code></pre>
</div>
<div id="visualize-differential-expression" class="section level2">
<h2>Visualize differential expression</h2>
<p>Visualize boxplots of a gene that is clearly differentially expressed (a) and one that is unclear due to increased variance (b).</p>
<pre class="r"><code>df_vis &lt;- data.frame(status = rep(c(&quot;con&quot;, &quot;treat&quot;), each = 50)) %&gt;%
  mutate(gene_de = c(rpois(n() / 2, lambda = 12), rpois(n() / 2, lambda = 30)),
         gene_var =c(rpois(n() / 2, lambda = 15) + rnorm(n() / 2, sd = 10),
                     rpois(n() / 2, lambda = 25) + rnorm(n() / 2, sd = 10)))
box_de &lt;- ggplot(df_vis, aes(x = status, y = gene_de)) +
  geom_boxplot() +
  theme_classic(base_size = 16) +
  ylim(0, 40) +
  labs(x = &quot;Treatment status&quot;, y = &quot;Gene expression level&quot;,
       title = &quot;Differential expression&quot;)
box_var &lt;- ggplot(df_vis, aes(x = status, y = gene_var)) +
  geom_boxplot() +
  theme_classic(base_size = 16) +
  ylim(0, 40) +
  labs(x = &quot;Treatment status&quot;, y = &quot;Gene expression level&quot;,
       title = &quot;High variance&quot;)
plot_grid(box_de, box_var, labels = letters[1:2])</code></pre>
<pre><code>Warning: Removed 1 rows containing non-finite values (stat_boxplot).</code></pre>
<pre><code>Warning: Removed 8 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="figure/linear-models.Rmd/boxplots-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-boxplots-1">
Past versions of boxplots-1.png
</button>
</p>
<div id="fig-boxplots-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/2372aa1e60c379d881b48e77a44fed980fc69c22/docs/figure/linear-models.Rmd/boxplots-1.png" target="_blank">2372aa1</a>
</td>
<td>
John Blischak
</td>
<td>
2019-01-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/49764900d3ce4ebe074d21c8736c0acaf1c94e97/docs/figure/linear-models.Rmd/boxplots-1.png" target="_blank">4976490</a>
</td>
<td>
John Blischak
</td>
<td>
2018-08-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="create-linear-model" class="section level2">
<h2>Create linear model</h2>
<p>As you just visualized, differential expression describes the situation in which a gene has a different mean expression level between conditions. While some gene expression patterns are easily diagnosed as differential expression or not from a quick visualization, you also saw some examples that were more ambiguous. Furthermore, you need a method that is more robust than a quick visual inspection and also scales to thousands of genes. For this you will use the tools of statistical inference to determine if the difference in mean expression level is larger than that expected by chance. Specifically, you will use linear models to perform the hypothesis tests. Linear models are an ideal choice for genomics experiments because their flexibility and robustness to assumptions allow you to conveniently model data from various study designs and data sources.</p>
<p>You should have already been introduced to linear models, for example in a DataCamp course such as <a href="https://www.datacamp.com/courses/correlation-and-regression">Correlation and Regression</a>, or in an introductory statistics course. Here I’ll review the terminology we will use in the remainder of the course, how to specify a linear model in R, and the intuition behind linear models.</p>
<p><span class="math display">\[ Y =  \beta_0 + \beta_1 X_1 + \epsilon \]</span></p>
<p>In this equation of a linear model, Y is the response variable. It must be a continuous variable. In the context of differential expression, it is a relative measure of either RNA or protein expression level for one gene. <span class="math inline">\(X_1\)</span> is an explanatory variable, which can be continuous or discrete, for example, control group versus treatment, or mutant versus wild type. <span class="math inline">\(\beta_1\)</span> quantifies the effect of the explanatory variable on the response variable. Furthermore, we can add additional explanatory variables to the equation for more complicated experimental designs. Lastly, models the random noise in the measurements.</p>
<p>In R, you specify a linear model with the function <code>lm</code>. This uses R’s formula syntax. On the left is the object that contains the response variable, and to the right of the tilde are the objects that contain the explanatory variables.</p>
<pre class="r"><code>lm(y ~ x1)</code></pre>
<p>A second explanatory variable can be added with a plus sign.</p>
<p><span class="math display">\[ Y =  \beta_0 + \beta_1 X_1 + \beta_2 X_2 + \epsilon \]</span></p>
<pre class="r"><code>lm(y ~ x1 + x2)</code></pre>
</div>
<div id="building-some-intuition-via-simulation" class="section level2">
<h2>Building some intuition via simulation</h2>
<p>The simulation below demonstrates how the statistical significance of the computed F-statistic of a linear model is equally affected by the noise (the residual sum of squares) and the signal (the explained sum of squares).</p>
<pre class="r"><code># Simulate a linear regression.
#
# n = sample size
# effect = slope
# error = standard deviation of distribution of residuals
# seed = seed for random number generator
#
# Returns a data.frame with the following columns:
#
# x          Explanatory variable
# y          Response variable
# y_bar      Mean of response variable
# intercept  Intercept of least squares regression line
# slope      Slope of least squares regression line
# y_hat      Fitted values
# fstat      F-statistic
# ss_exp     Explained sum of squares
# ss_res     Residual sum of squares (noise)
sim_lm &lt;- function(n, effect, error, seed = 1) {
  stopifnot(is.numeric(n), n &gt; 0, length(n) == 1)
  stopifnot(is.numeric(effect), length(effect) == 1)
  stopifnot(is.numeric(error), error &gt; 0, length(error) == 1)
  stopifnot(is.numeric(seed), length(seed) == 1)
  
  set.seed(seed)
  x = runif(n, min = -25, max = 25)
  y = x * effect + rnorm(n, sd = error)
  y_bar = mean(y)
  mod &lt;- lm(y ~ x)
  coefs &lt;- coef(mod)
  intercept &lt;- coefs[1]
  slope &lt;- coefs[2]
  y_hat = fitted(mod)
  anova_tidy &lt;- tidy(anova(mod))
  fstat &lt;- anova_tidy$statistic[1]
  ss &lt;- anova_tidy$sumsq
  ss_exp &lt;- ss[1]
  ss_res &lt;- ss[2]
  stopifnot(ss_exp - sum((y_hat - y_bar)^2) &lt; 0.01)
  stopifnot(ss_res - sum(residuals(mod)^2) &lt; 0.01)
  return(data.frame(x, y, y_bar, intercept, slope, y_hat, fstat, ss_exp, ss_res,
                    row.names = 1:n))
}

# Visualize the residual sum of squares
plot_ss_res &lt;- function(d) {
  ggplot(d, aes(x = x, y = y)) +
    geom_point() +
    geom_abline(aes(intercept = intercept, slope = slope)) +
    geom_linerange(aes(ymin = y, ymax = y_hat), color = &quot;red&quot;,
                   linetype = &quot;dashed&quot;) +
    geom_text(aes(x = quantile(x, 0.25), y = quantile(y, 0.75),
                  label = round(ss_res, 2)), color = &quot;red&quot;) +
    labs(title = &quot;Residual sum of squares (noise)&quot;) +
    ylim(-60, 60)
}

# Visualize the explained sum of squares
plot_ss_exp &lt;- function(d) {
  ggplot(d, aes(x = x, y = y)) +
    geom_abline(aes(intercept = intercept, slope = slope)) +
    geom_hline(aes(yintercept = y_bar)) +
    geom_linerange(aes(ymin = y_hat, ymax = y_bar), color = &quot;blue&quot;,
                   linetype = &quot;dashed&quot;) +
    geom_text(aes(x = quantile(x, 0.25), y = quantile(y, 0.75),
                  label = round(ss_exp, 2)), color = &quot;blue&quot;) +
    labs(title = &quot;Explained sum of squares&quot;) +
    ylim(-60, 60)
}</code></pre>
</div>
<div id="simulation-baseline" class="section level2">
<h2>Simulation: Baseline</h2>
<pre class="r"><code># baseline
baseline &lt;- sim_lm(n = 10, effect = 2, error = 5)
baseline_ss_res &lt;- plot_ss_res(baseline)
baseline_ss_exp &lt;- plot_ss_exp(baseline)
plot_grid(baseline_ss_res, baseline_ss_exp)</code></pre>
<p><img src="figure/linear-models.Rmd/baseline-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-baseline-1">
Past versions of baseline-1.png
</button>
</p>
<div id="fig-baseline-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/2372aa1e60c379d881b48e77a44fed980fc69c22/docs/figure/linear-models.Rmd/baseline-1.png" target="_blank">2372aa1</a>
</td>
<td>
John Blischak
</td>
<td>
2019-01-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/49764900d3ce4ebe074d21c8736c0acaf1c94e97/docs/figure/linear-models.Rmd/baseline-1.png" target="_blank">4976490</a>
</td>
<td>
John Blischak
</td>
<td>
2018-08-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>baseline$fstat[1]</code></pre>
<pre><code>[1] 280.3557</code></pre>
<p>The baseline simulation has an F-statistic of 280.3557372.</p>
</div>
<div id="simulation-increased-error" class="section level2">
<h2>Simulation: Increased error</h2>
<pre class="r"><code># Increased error
more_error &lt;- sim_lm(n = 10, effect = 2, error = 10)
more_error_ss_res &lt;- plot_ss_res(more_error)
more_error_ss_exp &lt;- plot_ss_exp(more_error)
plot_grid(more_error_ss_res, more_error_ss_exp)</code></pre>
<p><img src="figure/linear-models.Rmd/increased-error-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-increased-error-1">
Past versions of increased-error-1.png
</button>
</p>
<div id="fig-increased-error-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/2372aa1e60c379d881b48e77a44fed980fc69c22/docs/figure/linear-models.Rmd/increased-error-1.png" target="_blank">2372aa1</a>
</td>
<td>
John Blischak
</td>
<td>
2019-01-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/49764900d3ce4ebe074d21c8736c0acaf1c94e97/docs/figure/linear-models.Rmd/increased-error-1.png" target="_blank">4976490</a>
</td>
<td>
John Blischak
</td>
<td>
2018-08-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>more_error$fstat[1]</code></pre>
<pre><code>[1] 72.89125</code></pre>
<p>Doubling the error decreases the test statistic by a factor of 4.</p>
</div>
<div id="simulation-decreased-signal" class="section level2">
<h2>Simulation: Decreased signal</h2>
<pre class="r"><code># Decreased signal
less_signal &lt;- sim_lm(n = 10, effect = 1, error = 5)
less_signal_ss_res &lt;- plot_ss_res(less_signal)
less_signal_ss_exp &lt;- plot_ss_exp(less_signal)
plot_grid(less_signal_ss_res, less_signal_ss_exp)</code></pre>
<p><img src="figure/linear-models.Rmd/decreased-signal-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-decreased-signal-1">
Past versions of decreased-signal-1.png
</button>
</p>
<div id="fig-decreased-signal-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/2372aa1e60c379d881b48e77a44fed980fc69c22/docs/figure/linear-models.Rmd/decreased-signal-1.png" target="_blank">2372aa1</a>
</td>
<td>
John Blischak
</td>
<td>
2019-01-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/49764900d3ce4ebe074d21c8736c0acaf1c94e97/docs/figure/linear-models.Rmd/decreased-signal-1.png" target="_blank">4976490</a>
</td>
<td>
John Blischak
</td>
<td>
2018-08-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>less_signal$fstat[1]</code></pre>
<pre><code>[1] 72.89125</code></pre>
<p>Similarly, halving the signal decreases the test statistic by a factor of 4.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.3 (2019-03-11)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.2 LTS

Matrix products: default
BLAS: /usr/lib/x86_64-linux-gnu/atlas/libblas.so.3.10.3
LAPACK: /usr/lib/x86_64-linux-gnu/atlas/liblapack.so.3.10.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] knitr_1.22.6  dplyr_0.8.0.1 cowplot_0.9.4 ggplot2_3.1.0 broom_0.5.1  

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.1           pillar_1.3.1         compiler_3.5.3      
 [4] git2r_0.25.1         plyr_1.8.4           workflowr_1.2.0.9000
 [7] tools_3.5.3          digest_0.6.18        evaluate_0.13       
[10] tibble_2.1.1         nlme_3.1-137         gtable_0.3.0        
[13] lattice_0.20-38      pkgconfig_2.0.2      rlang_0.3.3         
[16] yaml_2.2.0           xfun_0.6             withr_2.1.2         
[19] stringr_1.4.0        generics_0.0.2       fs_1.2.7            
[22] rprojroot_1.2        grid_3.5.3           tidyselect_0.2.5    
[25] glue_1.3.1           R6_2.4.0             rmarkdown_1.12      
[28] purrr_0.3.2          tidyr_0.8.3          magrittr_1.5        
[31] whisker_0.3-2        backports_1.1.3      scales_1.0.0        
[34] htmltools_0.3.6      assertthat_0.2.1     colorspace_1.4-1    
[37] labeling_0.3         stringi_1.4.3        lazyeval_0.2.2      
[40] munsell_0.5.0        crayon_1.3.4        </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
